# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: gfx/thebes/gfxPlatformFontList.h
# Commit: cfdd1c5d8d3f
# Full Hash: cfdd1c5d8d3f6b99a6c6695dee91f12a154e16af
# Author: Jonathan Kew <jkew@mozilla.com>
# Date: 2021-09-08 21:39:05
# Regressor Bug: 1715501
# File Overlap Count: 1
# Description:
#   Bug 1715501 - patch 1 - Track codepoints with no available fonts and replacement-char family separately for each font-visibility level. r=emilio
#   
#   This does not change existing behavior, but will be required once font visibility
#   is no longer simply a global setting. The data cached in these members depends on
#   the font visibility level, so currently we just flush it if the visibility pref is
# ==============================================================================

diff -r 7518e48eae57 -r cfdd1c5d8d3f gfx/thebes/gfxPlatformFontList.h
--- a/gfx/thebes/gfxPlatformFontList.h	Wed Sep 08 09:33:17 2021 +0000
+++ b/gfx/thebes/gfxPlatformFontList.h	Wed Sep 08 09:35:16 2021 +0000
@@ -21,6 +21,7 @@
 
 #include "nsIMemoryReporter.h"
 #include "mozilla/Attributes.h"
+#include "mozilla/EnumeratedArray.h"
 #include "mozilla/FontPropertyTypes.h"
 #include "mozilla/MemoryReporting.h"
 #include "mozilla/Mutex.h"
@@ -509,7 +510,7 @@
       mozilla::StyleGenericFontFamily aGenericType);
 
   bool SkipFontFallbackForChar(uint32_t aCh) const {
-    return mCodepointsWithNoFonts.test(aCh);
+    return mCodepointsWithNoFonts[mVisibilityLevel].test(aCh);
   }
 
   // Return whether the given font-family record should be visible to CSS,
@@ -869,13 +870,16 @@
       mLangGroupPrefFonts;
   mozilla::UniquePtr<PrefFontList> mEmojiPrefFont;
 
-  // when system-wide font lookup fails for a character, cache it to skip future
-  // searches
-  gfxSparseBitSet mCodepointsWithNoFonts;
+  // When system-wide font lookup fails for a character, cache it to skip future
+  // searches. This is an array of bitsets, one for each FontVisibility level.
+  mozilla::EnumeratedArray<FontVisibility, FontVisibility::Count,
+                           gfxSparseBitSet>
+      mCodepointsWithNoFonts;
 
   // the family to use for U+FFFD fallback, to avoid expensive search every time
   // on pages with lots of problems
-  FontFamily mReplacementCharFallbackFamily;
+  mozilla::EnumeratedArray<FontVisibility, FontVisibility::Count, FontFamily>
+      mReplacementCharFallbackFamily;
 
   // Sorted array of lowercased family names; use ContainsSorted to test
   nsTArray<nsCString> mBadUnderlineFamilyNames;
